text/0431-sagemaker-l2-endpoint.md

```diff
@@ -0,0 +1,1150 @@
+     */
+    metricCPUUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Metric for memory utilization
+     *
+     * @default average over 5 minutes
+     */
+    metricMemoryUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Metric for GPU utilization
+     *
+     * @default average over 5 minutes
+     */
+    metricGPUUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Metric for GPU memory utilization
+     *
+     * @default average over 5 minutes
+     */
+    metricGPUMemoryUtilization(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
+    /**
+     * Enable autoscaling for SageMaker Endpoint production variant
+     *
+     * @param scalingProps EnableScalingProps
+     */
+    autoScaleInstanceCount(scalingProps: appscaling.EnableScalingProps): ScalableInstanceCount;
+  }
+
+  class EndpointProductionVariant implements IEndpointProductionVariant { ... }
+  ```
+
+### Is this a breaking change?
+
+No.
+
+### What alternative solutions did you consider?
+
+1. Theoretically, the `ContainerImage` code (referenced [above](#container-image)) from the
+   `@aws-cdk/ecs` and `@aws-cdk/sagemaker` modules could be unified assuming it would be sufficient
+   for both use-cases to `bind` using an `IGrantable` (adjusting ECS's `TaskDefinition`
+   accordingly). However, it's unclear within which module such a unified API should reside as
+   support for private repositories makes it a bad fit for `@aws-cdk/ecr` and `@aws-cdk/ecr-assets`,
+   and it would be unintuitive for `@aws-cdk/sagemaker` to declare a dependency on `@aws-cdk/ecs`.
+
+   Package concerns aside, historically, there was a period during which SageMaker only supported
+   ECR as an image source while ECS was capable of sourcing images from either ECR or a
+   customer-owned private repository. Given the fact that these two products' supported images
+   sources may yet again diverge in the future, maybe it would be best to keep their
+   `ContainerImage` APIs separate within their respective modules.
+1. In the [earliest PR][earliest-pr] attempting to add SageMaker L2 constructs to the CDK, the
+   author did not create an `EndpointConfig` construct, instead hiding the resource's creation
+   behind `Endpoint` (to which production variants could be added). Although a simplifier, this
+   prevents customers from reusing configuration across endpoints. For this reason, an explicit
+   L2 construct for endpoint configuration was incorporated into this RFC.
+
+[earliest-pr]: https://github.com/aws/aws-cdk/pull/2888
+
+### What are the drawbacks of this solution?
+
+This RFC and its [original associated implementation PR][original-pr] were based on a Q3 2019
+feature set of SageMaker real-time inference endpoints. Since that point in time, SageMaker has
+launched the following features which would require further additions to the L2 API contracts:
+
```

Comment:
If these are part of SageMaker's core feature set, I think they need to be taken into account in this RFP. We may be creating one way doors with contracts set here that will make the user experience more complicated if these are added later.

Reply:
At least I'd like a good story for how the API would evolve.

Reply:
Just to be clear. We won't actually be creating any one-way doors because sagemaker will start out as an alpha module. I think we should be thinking about how these constructs can be extended to support other SageMaker resources in the future, but it's not necessary to include them in the current RFC.

Reply:
@kaizencc I've familiarized myself with asynchronous inference and serverless inference (the two feature launches with the biggest impact on the proposed APIs in this RFC) by deploying a few sample endpoints via CloudFormation and found a number of feature disparities that I was hoping to get some design guidance on questions related to compile-time, synthesis-time, and deploy-time detection of issues. Particularly, strong compile-time seem like a good idea in practice, but in reality, I would hope that some of these feature gaps disappear. To avoid that future evolution from preventing convergence of the API contracts, using synthesis-time validation seems ideal, but then again, I have no guarantee that these products will actually converge. Anyway, to go into more detail:

1. **Lack of feature parity across inference products**: The [serverless inference documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/serverless-endpoints.html) states that the following features are absent from SageMaker's serverless feature:
    > Some of the features currently available for SageMaker Real-time Inference are not supported for Serverless Inference, including GPUs, AWS marketplace model packages, private Docker registries, Multi-Model Endpoints, VPC configuration, network isolation, data capture, multiple production variants, Model Monitor, and inference pipelines.

    Although I haven't found corresponding documentation, asynchronous inference has similar limitations: you cannot use multiple production variants nor can the single supported variant be serverless (it must used instance-based hosting). This led me down two separate paths:
   1. **Asynchronous Endpoints:** Whether or not an endpoint should be treated as sync vs async is controlled cross-variant at the level of the `EndpointConfig` resource. When an `EndpointConfig` resource is configured as async, it is currently only capable of supporting a single production variant. Should I use a compile-time solution (e.g. two different compile-time `*Props` interfaces to differentiate between the supported features of sync vs async) or instead simply use synthesis-time validation to identify a misconfigured resource (e.g., an async endpoint config with multiple variants)?
    1. **Serverless Production Variants:** Similarly, even though the instance-vs-serverless specification is per variant, the presence of one serverless production variant prevents customers from using additional variants (i.e., an aspect of one variant influences the behavior of the entire EndpointConfig resource). Would a synthesis-time check be sufficient here to error out on the incompatible variants or would a compile-time contract guarding this scenario be preferred?
1. **Feature-specific metrics:** Asynchronous endpoints have unique metrics like `ApproximateBacklogSize` on the endpoint name dimension whereas serverless variants have unique metrics like `ModelSetupTime` on the endpoint name + production variant dimension. When adding `metric*` helper APIs for these scenarios, should they be implemented in such a way to guarantee that the metric is available for the specific use-case from a compile-time perspective (i.e., each disparate use-case has an appropriate interface with only its suitable methods) or is it sufficient to use synthesis-time validation in this case?

Reply:
Hi @petermeansrock! These are all great questions, and thank you for doing the research on this.

`Asynchronous Endpoints`: I don't have a great answer. Sounds like `AsyncEndpointConfig` would be nice, but in the future, if `Async` supported multiple variants, suddenly`AsyncEndpointConfig` would be no different from `SyncEndpointConfig` and thus be redundant. Given this, I think it's safer to stay with one `EndpointConfig` construct with synth-time checks to make sure we catch misconfigured resources prior to deploy. They are not so different that having a generic `EndpointConfig` makes it awkward, so I think we're okay with keeping the two together.

`Serverless Production Variants`: Let me make sure I'm understanding right -- for the `productionVariant` property, there can be one or more serverless variants OR one or more instance variants. Just cannot have a mix of serverless and instance variants. In that case, I like the idea of having 2 properties, `serverlessProductionVariants` and `instanceProductionVariants` that are mutually exclusive (documented as such and validated at synth time).

`Feature-specific metrics`: I think the RFC should not worry about specific `XxxMetric` APIs and rather just expose the generic `metric` API and allow the user to specify exactly what metrics they want to use. This item can be a fast-follow item but isn't required for an alpha release, and there's already a lot of ideas to implement in this RFC. To answer your question though, I think synth-time validation makes sense / is the best we can do.

Feel free to override any of these decisions if you have a compelling reason -- just let me know why so I can be convinced :).

Reply:
@kaizencc 

> Given this, I think it's safer to stay with one `EndpointConfig` construct with synth-time checks to make sure we catch misconfigured resources prior to deploy. They are not so different that having a generic `EndpointConfig` makes it awkward, so I think we're okay with keeping the two together.

Sounds good, that makes perfect sense to me.

> `Serverless Production Variants`: Let me make sure I'm understanding right -- for the `productionVariant` property, there can be one or more serverless variants OR one or more instance variants. Just cannot have a mix of serverless and instance variants.

No, it's unfortunately even more limited:

1. If your `EndpointConfig` is asynchronous, you can have a single instance-based production variant.
2. For a synchronous `EndpointConfig`, you can either:
    1. Have a _single_ serverless production variant **OR**
    2. Have one or more instance-based production variants

In other words, multiple production variants are _only_ supported today for a synchronous `EndpointConfig` using instance-based hosting.

> `Feature-specific metrics`: I think the RFC should not worry about specific `XxxMetric` APIs and rather just expose the generic `metric` API and allow the user to specify exactly what metrics they want to use. This item can be a fast-follow item but isn't required for an alpha release, and there's already a lot of ideas to implement in this RFC. **To answer your question though, I think synth-time validation makes sense / is the best we can do.**

Sounds good!

Reply:
Thanks for clarifying. This does seem confusing to navigate, so its a great call out. The CDK can and should do the best it can to make things simpler for its users. The clarification has given me a change of heart, and now I'm interested in the following idea: two constructs, one generic `EndpointConfig`, and one "L2.5" `AsyncEndpointConfig` that provides further restrictions on the `productionVariant` property. My thinking is that users can drop down to use `EndpointConfig` for both sync and async endpoints if they want to, but we still have provided a higher level abstraction for async endpoints as well.

What do you think?

Reply:
@kaizencc 

> The clarification has given me a change of heart, and now I'm interested in the following idea: two constructs, one generic `EndpointConfig`, and one "L2.5" `AsyncEndpointConfig` that provides further restrictions on the productionVariant property.

## L2.5 `AsyncEndpointConfig` construct

For should such an `AsyncEndpointConfig`, which of the following would apply?
1. **`AsyncEndpointConfig extends Resource` and instantiates an `EndpointConfig` construct**: This is the classic L2 strategy, but would (I believe) introduce an intermediate async construct in the hierarchy, meaning customers couldn't replace an `AsyncEndpointConfig` with an `EndpointConfig` without changing their CloudFormation logical IDs.
1. **`AsyncEndpointConfig extends EndpointConfig`**: I assume this would retain the identical construct hierarchy but would violate the following [documented CDK design guideline](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md#construct-class):
    > As a rule of thumb, most constructs should directly extend the Construct or Resource instead of another construct. Prefer representing polymorphic behavior through interfaces and not through inheritance.
1. I'm assuming there could be another scenario where `AsyncEndpointConfig` instantiates `EndpointConfig` with a customer-provided scope and ID without extending `Construct`/`Resource` to avoid impacting the hierarchy, but I'm not familiar with this strategy being employed in the CDK codebase.

## Serverless Production Variants

The other limitation called out above, for synchronous endpoints, is that the presence of a single serverless variant prevents other variants from being configured. Would a synthesis-time approach be appropriate here (i.e., taking `ProductionVariant[]` on the `EndpointConfigProps` but erroring out if multiple are present alongside a serverless variant) or would you prefer a model more similar to your serverless-vs-instance specific properties?

Reply:
Thanks for reading the design guidelines! You are one of the few :). 

We don't have to implement `AsyncEndpointConfig` day 1 of the library. Lets add it to an "extensions" section and then turn those extensions into github issues later. 

I think this may be a special case where we are okay with extending `EndpointConfig`. We generally don't want to support that many L2.5s in the library, which is why we document that we want to extend the resource, but this could be a good candidate. I could be wrong, and be forced to change my mind later, but I'm looking at `NodeJsFunction` and other L2.5s in the main repo and they extend another construct. At any rate, this doesn't need to be implemented along with `EndpointConfig` so we don't need to finalize how to instantiate it just yet.

---

How about mutually exclusive properties `serverlessProductionVariant: ServerlessProductionVariant` and `productionVariants: ProductionVariant[]`? That seems like a good mix of property differentiation and synth time validation (that only one of the two props are set).

Reply:
@kaizencc 

> We don't have to implement `AsyncEndpointConfig` day 1 of the library. Lets add it to an "extensions" section and then turn those extensions into github issues later.

Sounds good! I assume the serverless pieces can be added to the same section.

> How about mutually exclusive properties `serverlessProductionVariant: ServerlessProductionVariant` and `productionVariants: ProductionVariant[]`? That seems like a good mix of property differentiation and synth time validation (that only one of the two props are set).

Two quick follow-up questions:

1. As both types of variants share a variant name & weight (and ideally could be used as siblings if the SageMaker folks release support for multi-heterogenous variants per endpoint config), to adjust your proposal slightly, would it make sense to define a shared interface like `ProductionVariant` and then have two interfaces extending it: `ServerlessProductionVariant` and `InstanceProductionVariant`? We could simply not export `ProductionVariant` yet as we don't have a need for polymorphic representation (e.g., `ProductionVariant[]`) and that way, both exported interfaces are specifically targeting their use-case (e.g., serverless vs instance-based).
2. Challenging one statement I made above: if the SageMaker folks later do support heterogeneous variants, would it make sense to collapse both variant specifications into a single `ProductionVariant[]` on the `EndpointConfigProps` (i.e., is that polymorphic API `jsii`-compatible)? If so (and assuming we can foresee SageMaker's product evolution), would it be unusual to present two separate variant specifications now if we _know_ they will eventually be "_doomed_" APIs (i.e., ideal for removal during a major version bump)?

***

Thanks so much for the help, Kaizen; this conversation has been immensely helpful! I'll start incorporating your recommended changes as my final pair of questions are mostly focused on minor remaining details that can be adjusted rather easily.

Reply:
Answer for 1 is yes. 

Answer for 2 is hmmm. One the one hand, predicting what SageMaker is going to do in the future is future-proofing and we want to avoid doing that. We have no guarantee that that will ever happen. If it happens when SageMaker is in alpha, then we can make that breaking change. If it happens after SageMaker is stabilized, then we can deprecate `serverlessProductionVariant` and `instanceProductionVariants` and introduce a new `productionVariant` property. Or, we can keep both APIs and just remove/modify the synth-time validation -- they're not necessarily _doomed_ and it might be fine to specify different production variants in different properties. I think if that happens it's okay, we're not stuck, and any changes we have to make won't really hurt the API too much.

Reply:
> `Feature-specific metrics`: I think the RFC should not worry about specific `XxxMetric` APIs and rather just expose the generic metric API and allow the user to specify exactly what metrics they want to use. This item can be a fast-follow item but isn't required for an alpha release, and there's already a lot of ideas to implement in this RFC. To answer your question though, I think synth-time validation makes sense / is the best we can do.

@kaizencc 

As I just published rev2, I wanted to highlight one bit: I kept the previously proposed `metric*` APIs on the newly renamed `IEndpointInstanceProductionVariant` as they had already been implemented and would be easy enough to remove if you feel strongly that they shouldn't be included.